### PR TITLE
Add regex to log retrieval.

### PR DESCRIPTION
**Pull request by BramGruneir**
Fetching logs from a file can now take a regex pattern to filter the log entries.

### CODE REVIEW

File: server/status.go [hunk 1]
```diff
@@ -243,11 +244,12 @@ func parseInt64WithDefault(s string, defaultValue int64) (int64, error) {
 
 // handleLocalLog returns the log entries parsed from the log file that occurred
 // after or on the query parameter "starttime" and before or on the query
-// parameter "endtime". Additionally, if a level is specified via the URL, the
-// log entries are scoped to that level. If no "starttime" is provided, then
-// a "starttime" of a day ago is used.  If no "endtime" is provided, the current
-// time is used. The cutoff for entries is defaults to defaultMaxLogEntries but
-// can be adjusted via the "max" query parameter.
+// parameter "endtime". If a query parameter "pattern" is provided, all results
+// must match the passed in regex. Additionally, if a level is specified via the
```
Comment (tamird) [inline server/status.go]: "must match" implies a caller responsibility

Comment (tamird) [inline server/status.go]: s/level/"level"/

File: server/status.go [hunk 2]
```diff
@@ -243,11 +244,12 @@ func parseInt64WithDefault(s string, defaultValue int64) (int64, error) {
...
+// URL, the log entries are scoped to that level. If no "starttime" is provided,
+// then a "starttime" of a day ago is used.  If no "endtime" is provided, the
+// current time is used. The cutoff for entries is defaults to
```
Comment (tamird) [inline server/status.go]: s/is //

Comment (tamird) [inline server/status.go]: I'd rewrite this comment in point form, personally

File: util/log/file.go [hunk 3]
```diff
@@ -292,10 +292,11 @@ func selectFiles(logFiles []FileInfo, severity Severity, endTimestamp int64) []F
 
 // FetchEntriesFromFiles fetches all available logs on disk that are of the
 // level of severity (or worse) and are between the start and end times. It will
```
Comment (tamird) [inline util/log/file.go]: that are of what level of severity?

File: util/log/file.go [hunk 4]
```diff
@@ -292,10 +292,11 @@ func selectFiles(logFiles []FileInfo, severity Severity, endTimestamp int64) []F
...
-// stop reading in new files if the maxEntries is exceeded. The logs entries
-// returned will be in decreasing order, with the closest log to the start
-// time being the first entry.
-func FetchEntriesFromFiles(severity Severity, startTimestamp, endTimestamp int64, maxEntries int) ([]LogEntry, error) {
+// stop reading in new files if the maxEntries is exceeded. Only log entries
```
Comment (tamird) [inline util/log/file.go]: s/the maxEntries/"maxEntries"/, s/is exceeded/is reached/

File: util/log/file.go [hunk 5]
```diff
@@ -292,10 +292,11 @@ func selectFiles(logFiles []FileInfo, severity Severity, endTimestamp int64) []F
...
+// that match the pattern will be output. The logs entries returned will be in
```
Comment (tamird) [inline util/log/file.go]: will be returned

File: server/status.go [hunk 6]
```diff
@@ -302,7 +304,16 @@ func (s *statusServer) handleLocalLog(w http.ResponseWriter, r *http.Request, ps
 		return
 	}
 
-	entries, err := log.FetchEntriesFromFiles(sev, startTimestamp, endTimestamp, int(maxEntries))
+	pattern := r.URL.Query().Get("pattern")
+	var regex *regexp.Regexp
+	if len(pattern) > 0 {
+		if regex, err = regexp.Compile(pattern); err != nil {
+			fmt.Fprintf(w, "could not compile regex pattern:%s", err)
```
Comment (spencerkimball) [inline server/status.go]: Shouldn't we be using log here as opposed to fmt.Fprintf?

File: util/log/file.go [hunk 7]
```diff
@@ -292,10 +292,11 @@ func selectFiles(logFiles []FileInfo, severity Severity, endTimestamp int64) []F
...
+// decreasing order, with the closest log to the start time being the first
```
Comment (tamird) [inline util/log/file.go]: decreasing order by what?

**Comment by spencerkimball** [comment]
LGTM

File: server/status.go (hunk 2 above)
Comment (BramGruneir) [inline server/status.go]: Yeah, it's too large, I'll do just that.

File: util/log/file.go [hunk 8]
```diff
@@ -349,7 +352,18 @@ func readAllEntriesFromFile(file FileInfo, startTimestamp, endTimestamp int64, m
...
 			return nil, false, err
 		}
-		if entry.Time >= startTimestamp && entry.Time <= endTimestamp {
+		var match bool
+		if pattern == nil {
+			match = true
+		} else {
+			args := []interface{}{}
+			for _, arg := range entry.Args {
+				args = append(args, arg.Str)
```
Comment (tamird) [inline util/log/file.go]: can you just check each fragment rather than building up a giant copy?

File: server/status.go (hunk 6 above)
Comment (BramGruneir) [inline server/status.go]: Fprintf to return the error to the caller.

Comment (tamird) [inline server/status.go]: does that mean you need an error code and a `return`, as well?

File: server/status_test.go [hunk 9]
```diff
@@ -428,6 +432,9 @@ func TestStatusLocalLogs(t *testing.T) {
 		if testCase.StartTimestamp > 0 {
 			fmt.Fprintf(&url, "endtime=%d&", testCase.EndTimestamp)
 		}
+		if len(testCase.Pattern) > 0 {
```
Comment (tamird) [inline server/status_test.go]: doesn't look like you're testing that the output is filtered?

File: server/status.go (hunk 1 above)
Comment (BramGruneir) [inline server/status.go]: fixed and reworded.

File: server/status.go (hunk 6 above)
Comment (BramGruneir) [inline server/status.go]: The return was there, but I was missing the error code.  Added.

File: server/status_test.go (hunk 9 above)
Comment (BramGruneir) [inline server/status_test.go]: I am indeed, with the last 3 test cases, they look for the pattern "Warning".

File: util/log/file.go (hunk 8 above)
Comment (BramGruneir) [inline util/log/file.go]: I wish I could.  But I can see someone looking for a pattern that bridges the format string and the values.
"failure processing range range=1" or something similar

File: util/log/file.go (hunk 4 above)
Comment (BramGruneir) [inline util/log/file.go]: Cleaned this whole comment up.

File: util/log/file.go (hunk 5 above)
Comment (BramGruneir) [inline util/log/file.go]: Cleaned this whole comment up.

File: util/log/file.go (hunk 7 above)
Comment (BramGruneir) [inline util/log/file.go]: Cleaned this whole comment up.

File: util/log/file.go [hunk 10]
```diff
@@ -327,12 +329,13 @@ func FetchEntriesFromFiles(severity Severity, startTimestamp, endTimestamp int64
...
-// occurred before the startTime to ensure inform the caller that no more log
-// files need to be processed. If the number of entries is exceeds maxEntries
-// then processing of new entries is stopped immediately.
-func readAllEntriesFromFile(file FileInfo, startTimestamp, endTimestamp int64, maxEntries int) ([]LogEntry, bool, error) {
+// between the start and end times and match the pattern if it exists. It
+// returns the entries in the reverse order, from newest to oldest. It also
+// returns a flag that denotes if any timestamp occurred before the startTime to
+// inform the caller that no more log files need to be processed. If the number
+// of entries is exceeds maxEntries then processing of new entries is stopped
+// immediately.
```
Comment (BramGruneir) [inline util/log/file.go]: Cleaned up this comment as well.

File: server/status.go [hunk 11]
```diff
@@ -241,13 +242,19 @@ func parseInt64WithDefault(s string, defaultValue int64) (int64, error) {
...
 
-// handleLocalLog returns the log entries parsed from the log file that occurred
-// after or on the query parameter "starttime" and before or on the query
-// parameter "endtime". Additionally, if a level is specified via the URL, the
-// log entries are scoped to that level. If no "starttime" is provided, then
-// a "starttime" of a day ago is used.  If no "endtime" is provided, the current
-// time is used. The cutoff for entries is defaults to defaultMaxLogEntries but
-// can be adjusted via the "max" query parameter.
+// handleLocalLog returns the log entries parsed from the log files stored on
+// the server. Log entries are returned in reverse chronological order with the
```
Comment (tamird) [inline server/status.go]: s/in reverse chronological order with the most recent being returned first/in descending chronological order/
the second half is redundant

File: server/status_test.go (hunk 9 above)
Comment (tamird) [inline server/status_test.go]: right, but where is the assertion that the returned logs actually match that pattern?

File: util/log/file.go [hunk 12]
```diff
@@ -290,12 +290,14 @@ func selectFiles(logFiles []FileInfo, severity Severity, endTimestamp int64) []F
...
-// level of severity (or worse) and are between the start and end times. It will
-// stop reading in new files if the maxEntries is exceeded. The logs entries
-// returned will be in decreasing order, with the closest log to the start
-// time being the first entry.
-func FetchEntriesFromFiles(severity Severity, startTimestamp, endTimestamp int64, maxEntries int) ([]LogEntry, error) {
+// FetchEntriesFromFiles fetches all available log entires on disk that match
+// the log 'severity' (or worse) and are between the 'startTimestamp' and
+// 'endTimestamp'. It will stop reading new files if the number of entries
+// exceeds 'maxEntries'. Log entries are further filtered by the regexp
+// 'pattern' if provided. The logs entries are returned in decreasing
```
Comment (tamird) [inline util/log/file.go]: same comment about redundant second half

File: util/log/file.go [hunk 13]
```diff
@@ -327,12 +330,13 @@ func FetchEntriesFromFiles(severity Severity, startTimestamp, endTimestamp int64
...
 
 // readAllEntriesFromFile reads in all log entries from a given file that are
-// between the start and end times and returns the entries in the reverse order,
-// from newest to oldest. It also returns a flag that denotes if any timestamp
-// occurred before the startTime to ensure inform the caller that no more log
-// files need to be processed. If the number of entries is exceeds maxEntries
-// then processing of new entries is stopped immediately.
-func readAllEntriesFromFile(file FileInfo, startTimestamp, endTimestamp int64, maxEntries int) ([]LogEntry, bool, error) {
+// between the 'startTimestamp' and 'endTimestamp' and match the 'pattern' if it
+// exists. It returns the entries in the reverse chronological order, from
```
Comment (tamird) [inline util/log/file.go]: ditto

**Comment by tamird** [comment]
LGTM modulo testing concern

**Comment by bdarnell** [comment]
LGTM

File: server/status.go (hunk 11 above)
Comment (BramGruneir) [inline server/status.go]: fixed it to just `reverse chronological order`
not sure how descending works in this context

File: util/log/file.go (hunk 12 above)
Comment (BramGruneir) [inline util/log/file.go]: fixed

File: util/log/file.go (hunk 13 above)
Comment (BramGruneir) [inline util/log/file.go]: fixed

**Pull request merged**